packages/framework/build.xml

```diff
@@ -1414,42 +1421,117 @@
         </trycatch>
     </target>
 
+    <target name="tests-check-failed">
+        <available file="${path.tests.failed}" type="file" property="tests.failed"/>
+        <if>
+            <isset property="tests.failed"/>
+            <then>
+                <phingcall target="tests-reset-failed"/>
+                <fail message="Tests failed"/>
+            </then>
+        </if>
+    </target>
+
     <target name="tests-functional" depends="tests-functional-application,tests-functional-frontend-api,tests-functional-frontend-api-b2b" description="Runs functional tests."/>
 
     <target name="tests-functional-application" depends="production-protection,clean,test-clean-redis,domains-info-load" description="Runs application functional tests." hidden="true">
-        <exec executable="${path.phpunit.executable}" logoutput="true" passthru="true" checkreturn="true">
+        <phingcall target="tests-reset-failed"/>
+        <foreach param="testsDirname" target="tests-functional-application-run">
+            <dirSet dir="${path.tests.functional}">
+                <include name="*/"/>
+                <exclude name="/"/>
+                <exclude name="*_*"/>
+                <exclude name="*/**/"/>
+            </dirSet>
+        </foreach>
+        <phingcall target="tests-check-failed"/>
+    </target>
+
+    <target name="tests-functional-application-run" description="Runs application functional tests in provided folder" hidden="true">
+        <echo message="Running tests from: ${path.tests.functional}${testsDirname}"/>
+        <exec executable="${path.phpunit.executable}" logoutput="true" passthru="true" checkreturn="${tests.stop-on-failed-group}" returnProperty="tests.exit-code">
             <arg value="--colors=always"/>
-            <arg value="--testsuite"/>
-            <arg value="Functional"/>
             <arg value="--exclude-group"/>
             <arg value="${phpunit-exclude-group}"/>
             <arg value="--configuration"/>
             <arg value="${path.root}/phpunit.xml"/>
+            <arg value="${path.tests.functional}${testsDirname}"/>
         </exec>
+        <if>
+            <not>
+                <equals arg1="${tests.exit-code}" arg2="0"/>
+            </not>
+            <then>
+                <phingcall target="tests-set-failed"/>
+            </then>
+        </if>
     </target>
 
     <target name="tests-functional-frontend-api" depends="production-protection,clean,test-clean-redis,domains-info-load" description="Runs Frontend API functional tests." hidden="true">
-        <exec executable="${path.phpunit.executable}" logoutput="true" passthru="true" checkreturn="true">
+        <phingcall target="tests-reset-failed"/>
+        <foreach param="testsDirname" target="tests-functional-frontend-api-run" inheritrefs="true" inheritall="true">
```

Comment:
are the `inheritrefs` and `inheritall` settings necessary? You do not define them in the `tests-functional-application` above :thinking: